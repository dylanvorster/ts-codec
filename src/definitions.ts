export type CommonMetadata = {
  description?: string;
  [key: string]: any;
};

export type CodecProps = {
  metadata?: CommonMetadata;
  required?: boolean;
  [key: string]: any;
};

export type Transformer<I, O> = (data: I) => O;

export type Codec<I, O, T = string, P = CodecProps> = {
  _tag: T;
  props: CodecProps & P;

  meta: (metadata: CommonMetadata) => Codec<I, O, T, P>;

  encode: Transformer<I, O>;
  decode: Transformer<O, I>;

  and: <T extends AnyCodec>(codec: T) => Intersection<Codec<I, O>, T>;
  or: <T extends AnyCodec>(codec: T) => Union<Codec<I, O>, T>;

  optional: () => OptionalCodec<Codec<I, O, T, P>>;
};

export type PassThroughCodec<T> = Codec<T, T>;

export type AnyCodec = Codec<any, any, any>;

export type Cx<C extends AnyCodec> = C extends Codec<infer I, infer O, infer T, infer P>
  ? { I: I; O: O; T: T; P: P }
  : never;
export type Ix<C extends AnyCodec> = Cx<C>['I'];
export type Ox<C extends AnyCodec> = Cx<C>['O'];

export type Encoded<T extends AnyCodec> = Ox<T>;
export type Decoded<T extends AnyCodec> = Ix<T>;

export type AnyObjectCodecShape = Record<string, AnyCodec>;

type OptionalKeys<T extends object> = {
  [K in keyof T]: undefined extends T[K] ? K : never;
}[keyof T];

type RequiredKeys<T extends object> = Exclude<keyof T, OptionalKeys<T>>;

type WithOptional<T extends object> = {
  [K in RequiredKeys<T>]: T[K];
} & {
  [K in OptionalKeys<T>]?: T[K];
};

type MappedIx<T extends AnyObjectCodecShape> = WithOptional<{ [K in keyof T]: Ix<T[K]> }>;
type MappedOx<T extends AnyObjectCodecShape> = WithOptional<{ [K in keyof T]: Ox<T[K]> }>;

export enum CodecType {
  String = 'string',
  Number = 'number',
  Boolean = 'boolean',
  Literal = 'literal',
  Enum = 'enum',
  Null = 'null',
  Any = 'any',

  Object = 'object',
  Record = 'record',
  Array = 'array',
  Tuple = 'tuple',

  Recursive = 'recursive',

  Union = 'union',
  Intersection = 'intersection'
}

export type Intersection<C1 extends AnyCodec, C2 extends AnyCodec> = Codec<
  Ix<C1> & Ix<C2>,
  Ox<C1> & Ox<C2>,
  CodecType.Intersection,
  {
    codecs: AnyCodec[];
  }
>;

export type Union<C1 extends AnyCodec, C2 extends AnyCodec> = Codec<
  Ix<C1> | Ix<C2>,
  Ox<C1> | Ox<C2>,
  CodecType.Union,
  {
    codecs: AnyCodec[];
  }
>;

export type ObjectCodec<T extends AnyObjectCodecShape> = Codec<
  MappedIx<T>,
  MappedOx<T>,
  CodecType.Object,
  {
    shape: T;
  }
>;

export type RecordCodec<T extends AnyCodec> = Codec<
  Record<string, Ix<T>>,
  Record<string, Ox<T>>,
  CodecType.Record,
  {
    type: T;
  }
>;

export type ArrayCodec<T extends AnyCodec> = Codec<
  Ix<T>[],
  Ox<T>[],
  CodecType.Array,
  {
    type: T;
  }
>;

export type AnyTuple = [AnyCodec, ...AnyCodec[]];

export type AssertArray<T> = T extends any[] ? T : never;
export type TupleIx<T extends AnyTuple> = AssertArray<{
  [k in keyof T]: T[k] extends AnyCodec ? Ix<T[k]> : never;
}>;
export type TupleOx<T extends AnyTuple> = AssertArray<{
  [k in keyof T]: T[k] extends AnyCodec ? Ox<T[k]> : never;
}>;

export type TupleCodec<T extends AnyTuple> = Codec<
  TupleIx<T>,
  TupleOx<T>,
  CodecType.Tuple,
  {
    codecs: T;
  }
>;

export type RecursiveCodec<T extends AnyCodec> = Codec<
  Ix<T>,
  Ox<T>,
  CodecType.Recursive,
  {
    id: string;
    resolver: () => T;
  }
>;

export type OptionalCodec<T extends AnyCodec> = Codec<Ix<T> | undefined, Ox<T> | undefined, Cx<T>['T'], Cx<T>['P']>;

type IdentityMapping<T extends CodecType> = T extends CodecType.String
  ? string
  : T extends CodecType.Number
  ? number
  : T extends CodecType.Boolean
  ? boolean
  : never;

export type IdentityCodec<T extends CodecType> = Codec<IdentityMapping<T>, IdentityMapping<T>, T>;

export type LiteralCodec<T extends string> = Codec<
  T,
  T,
  CodecType.Literal,
  {
    value: T;
  }
>;

export type EnumLike = {
  [key: string]: string;
};
export type EnumCodec<T extends EnumLike> = Codec<
  T[keyof T],
  T[keyof T],
  CodecType.Enum,
  {
    enum: EnumLike;
  }
>;
