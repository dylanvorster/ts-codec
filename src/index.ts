export * from './definitions';

export * from './types/primitives';
export * from './types/recursive';
export * from './types/arrays';
export * from './types/codec';
export * from './types/maps';

export * from './utils';
