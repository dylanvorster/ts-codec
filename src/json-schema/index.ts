export * from './definitions';
export * from './generator';
export * from './parsers';
export * from './utils';
