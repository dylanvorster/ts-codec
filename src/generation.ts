import { AnyCodec, CodecType, IdentityCodec } from './definitions';
import * as defs from './definitions';

export enum TransformTarget {
  Encoded = 'encoded',
  Decoded = 'decoded'
}

export type CommonParams = {
  target: TransformTarget;
  parsers: Parser<string, AnyCodec>[];
  allowAdditional: boolean;
};

export type GenerationOptions = Partial<CommonParams>;

type RecursiveCodecCache = Map<string, any>;
type GenerationContext = CommonParams & {
  cache: RecursiveCodecCache;
};

export type ParserFunction<C extends AnyCodec> = (codec: C, options: GenerationContext) => object;
export type Parser<T extends string, C extends AnyCodec> = {
  tag: T;
  parse: ParserFunction<C>;
};

export const createParser = <C extends AnyCodec, T extends string = string>(
  tag: T,
  parser: ParserFunction<C>
): Parser<T, C> => {
  return {
    tag,
    parse: parser
  };
};

const createPrimitiveParser = <T extends CodecType>(type: T): Parser<T, IdentityCodec<T>> => {
  return {
    tag: type,
    parse: () => {
      return { type: type };
    }
  };
};

const StringParser = createPrimitiveParser(CodecType.String);
const NumberParser = createPrimitiveParser(CodecType.Number);
const BooleanParser = createPrimitiveParser(CodecType.Boolean);
const NullParser = createPrimitiveParser(CodecType.Null);

const AnyParser = createParser<defs.AnyCodec>(CodecType.Any, (codec) => {
  return {};
});

const EnumParser = createParser<defs.EnumCodec<any>>(CodecType.Enum, (codec) => {
  return {
    type: 'string',
    enum: Object.values(codec.props.enum)
  };
});

const LiteralParser = createParser<defs.LiteralCodec<any>>(CodecType.Literal, (codec) => {
  return {
    type: 'string',
    const: codec.props.value
  };
});

const ObjectParser = createParser<defs.ObjectCodec<defs.AnyObjectCodecShape>>(CodecType.Object, (codec, options) => {
  const entries = Object.entries(codec.props.shape);
  return {
    type: 'object',
    properties: entries.reduce((acc: Record<string, any>, [key, codec]) => {
      acc[key] = RootParser(codec, options);
      return acc;
    }, {}),
    additionalProperties: !!options?.allowAdditional,
    required: entries
      .filter(([, codec]) => {
        return codec.props.required;
      })
      .map(([key]) => key)
  };
});

const RecordParser = createParser<defs.RecordCodec<defs.AnyCodec>>(CodecType.Record, (codec, options) => {
  return {
    type: 'object',
    additionalProperties: RootParser(codec.props.type, options),
    properties: {},
    required: []
  };
});

const ArrayParser = createParser<defs.ArrayCodec<AnyCodec>>(CodecType.Array, (codec, options) => {
  return {
    type: 'array',
    items: RootParser(codec.props.type, options)
  };
});

const TupleParser = createParser<defs.TupleCodec<defs.AnyTuple>>(CodecType.Tuple, (codec, options) => {
  return {
    type: 'array',
    items: codec.props.codecs.map((codec) => RootParser(codec, options)),
    minItems: codec.props.codecs.length,
    maxItems: codec.props.codecs.length
  };
});

const IntersectionParser = createParser<defs.Intersection<AnyCodec, AnyCodec>>(
  CodecType.Intersection,
  (codec, options) => {
    const schemas = codec.props.codecs.map((codec) => RootParser(codec, options));

    // return {
    //   type: "object",
    //   properties: schemas.reduce((properties, schema: any) => {
    //     return {
    //       ...properties,
    //       ...(schema.properties || {}),
    //     }
    //   }, {}),
    //   additionalProperties: !!options?.allowAdditional,
    //   required: schemas.reduce((required: string[], schema: any) => {
    //     return required.concat(schema.required || [])
    //   }, []),
    // }

    return {
      allOf: schemas
    };
  }
);

const UnionParser = createParser<defs.Union<AnyCodec, AnyCodec>>(CodecType.Union, (codec, options) => {
  return {
    anyOf: codec.props.codecs.map((codec) => RootParser(codec, options))
  };
});

const RecursiveParser = createParser<defs.RecursiveCodec<AnyCodec>>(CodecType.Recursive, (codec, options) => {
  const ref = {
    $ref: `#/definitions/${codec.props.id}`
  };

  const cached = options.cache.get(codec.props.id);
  if (cached) {
    return ref;
  }

  options.cache.set(codec.props.id, {});
  /**
   * The schema is applied to the definition _after_ pushing the definition to the cache
   * to ensure that there are no cache misses from calling RootParser
   */
  options.cache.set(codec.props.id, RootParser(codec.props.resolver(), options));

  return ref;
});

const RootParser = (codec: AnyCodec, options: GenerationContext) => {
  const parser = options.parsers.find((parser) => parser.tag === codec._tag);
  if (!parser) {
    throw new Error(`No parser configured for codec ${codec._tag}`);
  }
  const schema = parser.parse(codec, options);

  if (codec.props.metadata?.description) {
    return {
      description: codec.props.metadata.description,
      ...schema
    };
  }

  return schema;
};

export const generateJSONSchema = (codec: AnyCodec, options?: GenerationOptions): Record<string, any> => {
  const parsers: Parser<any, any>[] = [
    ...(options?.parsers || []),
    AnyParser,
    StringParser,
    NumberParser,
    BooleanParser,
    NullParser,
    LiteralParser,
    EnumParser,
    ObjectParser,
    RecordParser,
    ArrayParser,
    TupleParser,
    IntersectionParser,
    UnionParser,
    RecursiveParser
  ];

  const recursion_cache: RecursiveCodecCache = new Map();
  const schema = RootParser(codec, {
    parsers,
    target: options?.target ?? TransformTarget.Encoded,
    allowAdditional: options?.allowAdditional ?? false,
    cache: recursion_cache
  });

  return {
    definitions: Object.fromEntries(recursion_cache.entries()),
    ...schema
  };
};
