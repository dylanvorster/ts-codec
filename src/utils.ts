import * as defs from './definitions';

export class TransformError extends Error {
  constructor(readonly errors: string[]) {
    super(errors.join(', '));
    this.name = 'TransformError';
  }
}

export function isCodecType<T extends defs.CodecType.String | defs.CodecType.Number | defs.CodecType.Boolean>(
  codec: defs.AnyCodec,
  type: T
): codec is defs.IdentityCodec<T>;
export function isCodecType(codec: defs.AnyCodec, type: defs.CodecType.Object): codec is defs.ObjectCodec<any>;
export function isCodecType(
  codec: defs.AnyCodec,
  type: defs.CodecType.Intersection
): codec is defs.Intersection<defs.AnyCodec, defs.AnyCodec>;
export function isCodecType(
  codec: defs.AnyCodec,
  type: defs.CodecType.Union
): codec is defs.Union<defs.AnyCodec, defs.AnyCodec>;
export function isCodecType(codec: defs.AnyCodec, type: defs.CodecType) {
  return codec._tag === type;
}
