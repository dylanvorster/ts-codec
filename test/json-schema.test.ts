import * as t from '../src';

describe('json-schema', () => {
  const date = t.codec<Date, string>(
    'Date',
    (date) => date.toISOString(),
    (date) => new Date(date)
  );

  test('it should handle a simple identity schema', () => {
    const identity_schema = t.object({
      string: t.string,
      number: t.number,
      boolean: t.boolean,
      multi: t.string.or(t.number).or(t.boolean),
      optional: t.string.optional(),

      array: t.array(t.string),
      deeper_array: t.array(t.object({ a: t.string })),

      record: t.record(t.string),

      intersection: t
        .object({
          a: t.string
        })
        .and(
          t.object({
            b: t.number
          })
        )
    });

    expect(t.generateJSONSchema(identity_schema)).toMatchSnapshot();
  });

  test('it should generate schemas from dual codecs', () => {
    const schema = t.object({
      created_at: date
    });

    const DateParser = t.createParser('Date', (codec, options) => {
      switch (options.target) {
        case t.TransformTarget.Encoded: {
          return {
            type: 'string'
          };
        }

        case t.TransformTarget.Decoded: {
          return {
            nodeType: 'date'
          };
        }
      }
    });

    expect(
      t.generateJSONSchema(schema, { target: t.TransformTarget.Encoded, parsers: [DateParser] })
    ).toMatchSnapshot();
    expect(
      t.generateJSONSchema(schema, { target: t.TransformTarget.Decoded, parsers: [DateParser] })
    ).toMatchSnapshot();
  });

  test('it should generate schemas from recursive codecs', () => {
    type Schema = {
      a: string;
      b?: Schema;
    };

    const schema: t.Codec<Schema, Schema> = t.recursive(() => {
      return t.object({
        a: t.string,
        b: schema.optional()
      });
    });

    expect(t.generateJSONSchema(schema)).toMatchSnapshot();
  });
});
