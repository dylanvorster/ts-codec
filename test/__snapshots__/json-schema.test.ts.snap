// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json-schema it should generate schemas from dual codecs 1`] = `
Object {
  "additionalProperties": false,
  "definitions": Object {},
  "properties": Object {
    "created_at": Object {
      "type": "string",
    },
  },
  "required": Array [
    "created_at",
  ],
  "type": "object",
}
`;

exports[`json-schema it should generate schemas from dual codecs 2`] = `
Object {
  "additionalProperties": false,
  "definitions": Object {},
  "properties": Object {
    "created_at": Object {
      "nodeType": "date",
    },
  },
  "required": Array [
    "created_at",
  ],
  "type": "object",
}
`;

exports[`json-schema it should generate schemas from recursive codecs 1`] = `
Object {
  "$ref": "#/definitions/RecursiveCodec",
  "definitions": Object {
    "RecursiveCodec": Object {
      "additionalProperties": false,
      "properties": Object {
        "a": Object {
          "type": "string",
        },
        "b": Object {
          "$ref": "#/definitions/RecursiveCodec",
        },
      },
      "required": Array [
        "a",
      ],
      "type": "object",
    },
  },
}
`;

exports[`json-schema it should handle a simple identity schema 1`] = `
Object {
  "additionalProperties": false,
  "definitions": Object {},
  "properties": Object {
    "any": Object {},
    "array": Object {
      "items": Object {
        "type": "string",
      },
      "type": "array",
    },
    "boolean": Object {
      "type": "boolean",
    },
    "deeper_array": Object {
      "items": Object {
        "additionalProperties": false,
        "properties": Object {
          "a": Object {
            "type": "string",
          },
        },
        "required": Array [
          "a",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "enum": Object {
      "enum": Array [
        "a",
        "b",
      ],
      "type": "string",
    },
    "intersection": Object {
      "anyOf": Array [
        Object {
          "additionalProperties": false,
          "properties": Object {
            "a": Object {
              "type": "string",
            },
            "b": Object {
              "type": "number",
            },
          },
          "required": Array [
            "a",
            "b",
          ],
          "type": "object",
        },
        Object {
          "additionalProperties": false,
          "properties": Object {
            "a": Object {
              "type": "string",
            },
            "c": Object {
              "type": "number",
            },
          },
          "required": Array [
            "a",
            "c",
          ],
          "type": "object",
        },
      ],
    },
    "literal": Object {
      "const": "literal",
      "type": "string",
    },
    "multi": Object {
      "anyOf": Array [
        Object {
          "type": "string",
        },
        Object {
          "type": "number",
        },
        Object {
          "type": "boolean",
        },
      ],
    },
    "null": Object {
      "type": "null",
    },
    "number": Object {
      "type": "number",
    },
    "optional": Object {
      "type": "string",
    },
    "record": Object {
      "additionalProperties": Object {
        "type": "string",
      },
      "properties": Object {},
      "required": Array [],
      "type": "object",
    },
    "string": Object {
      "type": "string",
    },
    "tuple": Object {
      "items": Array [
        Object {
          "type": "string",
        },
        Object {
          "type": "number",
        },
      ],
      "maxItems": 2,
      "minItems": 2,
      "type": "array",
    },
    "with_description": Object {
      "description": "This is a description",
      "type": "string",
    },
  },
  "required": Array [
    "string",
    "number",
    "boolean",
    "any",
    "null",
    "literal",
    "enum",
    "tuple",
    "multi",
    "array",
    "deeper_array",
    "record",
    "intersection",
    "with_description",
  ],
  "type": "object",
}
`;

exports[`json-schema it should intersect a set of only unions 1`] = `
Object {
  "allOf": Array [
    Object {
      "anyOf": Array [
        Object {
          "additionalProperties": false,
          "properties": Object {
            "a": Object {
              "type": "string",
            },
          },
          "required": Array [
            "a",
          ],
          "type": "object",
        },
        Object {
          "additionalProperties": false,
          "properties": Object {
            "b": Object {
              "type": "string",
            },
          },
          "required": Array [
            "b",
          ],
          "type": "object",
        },
      ],
    },
    Object {
      "anyOf": Array [
        Object {
          "additionalProperties": false,
          "properties": Object {
            "c": Object {
              "type": "string",
            },
          },
          "required": Array [
            "c",
          ],
          "type": "object",
        },
        Object {
          "additionalProperties": false,
          "properties": Object {
            "d": Object {
              "type": "string",
            },
          },
          "required": Array [
            "d",
          ],
          "type": "object",
        },
      ],
    },
  ],
  "definitions": Object {},
}
`;

exports[`json-schema it should merge intersections 1`] = `
Object {
  "additionalProperties": false,
  "definitions": Object {},
  "properties": Object {
    "a": Object {
      "type": "string",
    },
    "b": Object {
      "type": "string",
    },
    "c": Object {
      "type": "string",
    },
  },
  "required": Array [
    "a",
    "b",
    "c",
  ],
  "type": "object",
}
`;

exports[`json-schema it should merge intersections 2`] = `
Object {
  "additionalProperties": false,
  "definitions": Object {},
  "properties": Object {
    "a": Object {
      "type": "string",
    },
  },
  "required": Array [
    "a",
  ],
  "type": "object",
}
`;

exports[`json-schema it should product intersections containing union and object schemas 1`] = `
Object {
  "anyOf": Array [
    Object {
      "additionalProperties": false,
      "properties": Object {
        "a": Object {
          "type": "string",
        },
        "b": Object {
          "type": "string",
        },
      },
      "required": Array [
        "a",
        "b",
      ],
      "type": "object",
    },
    Object {
      "additionalProperties": false,
      "properties": Object {
        "a": Object {
          "type": "string",
        },
        "c": Object {
          "type": "string",
        },
      },
      "required": Array [
        "a",
        "c",
      ],
      "type": "object",
    },
  ],
  "definitions": Object {},
}
`;
